### ISSUE TITLE
yield violates method return type

### ISSUE THREAD

**Issue by al6x** [crystal-lang/crystal#11033]
## Bug Report

``` Crystal
def twice : Int
  yield 1
  yield 2
end

p twice { |i| i + 1 }         # => 3
p twice { |i| break "hello" } # => "hello"
```

```
crystal --version
Crystal 1.1.1 [6d9a1d583] (2021-07-26)

LLVM: 10.0.0
Default target: x86_64-apple-macosx
```

**Comment by lbguilherme** [comment]
My interpretation of how `break` works is that it "aborts the execution of the method, with a value", instead of saying it "makes the method return early, with a value". I don't think the behavior should change.

That said, the [documentation](https://crystal-lang.org/reference/syntax_and_semantics/blocks_and_procs.html#break) needs clarification because it says the exact opposite:

> A `break` expression inside a block exits early from the method. [...] `break` can also accept arguments: these become the method's return value.

**Comment by asterite** [comment]
The way `break` for methods is coded right now is a bit weird, I noticed this while coding the interpreter. There's a special "break" node that gets all types, instead of binding directly to the method's return type. If that's implemented correctly, it would have caught that.

So I think this is a legit bug, though I agree that `break` is kind of outside of the method's logic. So for now there's no real pressure to change this.

**Comment by beta-ziliani** [comment]
From a practical PoV, I think it's good to avoid the developer shoot in their foot, so it's best to have this fixed by some definition of "soon". In particular, as it might break existing code (but it's a bug).

**Comment by straight-shoota** [comment]
Thanks, @stakach for taking the initiative on https://github.com/crystal-lang/crystal/pull/16851 🫶 

Unfortunately, I don't think there is an easy path forward because existing codebases (including the standard library) expect the current buggy behaviour.

So we'll have to change this gradually for a smooth transition because a compiler error would be a breaking change.

**Comment by straight-shoota** [comment]
On the other hand: following the argument from https://github.com/crystal-lang/crystal/issues/11033#issuecomment-888217542 and https://github.com/crystal-lang/crystal/issues/11033#issuecomment-888264155, `break` behaviour can be considered outside the method's scope because it's actually defined by the caller.

I was thinking about how we could express a return type that includes values from `break` in the method signature. Maybe something like `foo(&) : Nil | break`. But what would be the point?

Technically, any method that accepts a block argument can return any arbitrary type with a `break` in the block.
We could just accept this as implicit behaviour which does not need explicit documentation for every single yielding method (or make this implicit behaviour automatically explicit in the doc generator).

That would implicitly make the current state _correct_. And I think that might be entirely fine.

**Comment by stakach** [comment]
I agree with you @straight-shoota - the current behavior of break seems fine to me.
This issue has been open for years with little complaint.
Happy for you to close both this and my pull request if the plan is to leave things as-is

**Comment by straight-shoota** [comment]
Closing. This works as expected.

The return type restriction of a method describes which types it can return. But when breaking from a block, the method technically does not _return_ because the caller interrupts its execution.
